File: platform/core-impl/src/com/intellij/ide/plugins/ModuleDependenciesDescriptor.java
// Copyright 2000-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.ide.plugins;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.List;

final class ModuleDependenciesDescriptor {
  static final ModuleDependenciesDescriptor EMPTY = new ModuleDependenciesDescriptor(Collections.emptyList());

  final List<ModuleItem> modules;

  ModuleDependenciesDescriptor(@NotNull List<ModuleItem> modules) {
    this.modules = modules;
  }

  @Nullable ModuleDependenciesDescriptor.ModuleItem findModuleByName(@NotNull String name) {
    for (ModuleDependenciesDescriptor.ModuleItem module : modules) {
      if (module.name.equals(name)) {
        return module;
      }
    }
    return null;
  }

  static final class ModuleItem {
    final String name;
    final String packageName;

    ModuleItem(@NotNull String name, @Nullable String packageName) {
      this.name = name;
      this.packageName = packageName;
      if (packageName != null && packageName.endsWith(".")) {
        throw new RuntimeException("packageName must not ends with dot: " + packageName);
      }
    }
  }
}

File: platform/core-impl/src/com/intellij/ide/plugins/ModuleDependenciesDescriptor.kt
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.ide.plugins

import com.intellij.openapi.extensions.PluginId
import com.intellij.util.Java11Shim
import org.jetbrains.annotations.ApiStatus
import java.util.*

@ApiStatus.Internal
class ModuleDependenciesDescriptor(@JvmField val modules: List<ModuleReference>, @JvmField val plugins: List<PluginReference>) {
  companion object {
    @JvmField val EMPTY: ModuleDependenciesDescriptor = ModuleDependenciesDescriptor(Collections.emptyList(), Collections.emptyList())
  }

  class ModuleReference(@JvmField val name: String) {
    override fun toString(): String = "ModuleItem(name=$name)"
  }

  class PluginReference(@JvmField val id: PluginId) {
    override fun toString(): String = "PluginReference(id=$id)"
  }

  override fun toString(): String = "ModuleDependenciesDescriptor(modules=$modules, plugins=$plugins)"
}

@ApiStatus.Internal
class PluginContentDescriptor(@JvmField val modules: List<ModuleItem>) {
  companion object {
    @JvmField val EMPTY: PluginContentDescriptor = PluginContentDescriptor(Java11Shim.INSTANCE.listOf())
  }

  @ApiStatus.Internal
  class ModuleItem(
    @JvmField val name: String,
    @JvmField val configFile: String?,
    @JvmField internal val descriptorContent: CharArray?,
    @JvmField val loadingRule: ModuleLoadingRule,
  ) {
    @JvmField
    internal var descriptor: IdeaPluginDescriptorImpl? = null

    fun requireDescriptor(): IdeaPluginDescriptorImpl = descriptor ?: throw IllegalStateException("Descriptor is not set for $this")

    override fun toString(): String = "ModuleItem(name=$name, descriptor=$descriptor, configFile=$configFile)"
  }

  override fun toString(): String = "PluginContentDescriptor(modules=$modules)"
}